csrc/device_lower/pass/predicate.cpp

```diff
@@ -89,11 +97,100 @@ class ConditionalFromPredicateModifier : public kir::ExprMutator {
       expr->predicate()->setValue(conditional);
       NVF_ERROR(expr->predicate()->value() != nullptr);
       setWritePredicate(expr);
+    } else if (
+        expr && expr->isA<kir::MBarrierWaitParity>() && ublk_predicate_val_) {
+      auto fl = for_loops_.front();
+      std::unordered_map<Val*, Val*> replace_map;
+      replace_map[tma_branch_fl_index_] = fl->index();
+      auto local_predicate_val =
+          ir_utils::replaceValRecursively(ublk_predicate_val_, replace_map);
+      local_predicate_val = GpuLower::current()->commonScalarMap().hoistScalar(
+          local_predicate_val, {for_loops_.front()});
+      kir::Predicate* pred =
+          IrBuilder::create<kir::Predicate>(local_predicate_val);
+      kir::IfThenElse* inline_ite = IrBuilder::create<kir::IfThenElse>(pred);
+      kir::ExprMutator::registerReplace(expr, inline_ite);
+      inline_ite->thenBody().push_back(expr);
     }
 
     kir::ExprMutator::dispatch(expr);
   }
 
+  // Move Inline predicate for UBLK TMA load to ElectSync
+  // (1) UBLK TMA load can't handle out-of-bound access, it has Inline
+  //     predicates added in Unroll pass.
+  // (2) TMA loads are synced using MBarrierArriveExpectTx and
+  //     MBarrierWaitParity, they need the same predicate to avoid deadlock.
+  // Returns false if expr is not an IfThenElse with Inline predicate.
+  bool reviseUblkPredicate(Expr* expr, Val* conditional) {
+    // Looking for the following pattern:
+    // IF Inline
+    //   Ts = UBLK TMA load
+    auto ite = expr->as<kir::IfThenElse>();
+    if (ite == nullptr) {
+      return false;
+    }
+    if (ite->predicate()->predicate_type() != PredicateType::Inline) {
+      return false;
+    }
+    for (auto lexpr : ite->thenBody().exprs()) {
+      if (!ir_utils::isCpAsyncUblk(lexpr)) {
+        continue;
+      }
+      // Ublk predicate is only used for warp specialized normalization kernel,
+      // it assumes the outer dim is scheduled as [I/Unroll/BIDy, BIDy,
```

Comment:
There's a way to add a condition that should be validated at runtime. See `GpuLower::validate`.